samza-core/src/main/java/org/apache/samza/control/WatermarkManager.java

```diff
@@ -0,0 +1,191 @@
+   *   <li>Watermark(stream) = min { Watermark(ssp) | ssp is a partition of stream that assigns to this task } </li>
+   * </ul>
+   *
+   * @param envelope the envelope contains {@link WatermarkMessage}
+   * @return watermark envelope if there is a new aggregate watermark for the stream
+   */
+  public Watermark update(IncomingMessageEnvelope envelope) {
+    SystemStreamPartition ssp = envelope.getSystemStreamPartition();
+    WatermarkState state = watermarkStates.get(ssp);
+    WatermarkMessage message = (WatermarkMessage) envelope.getMessage();
+    state.update(message.getTimestamp(), message.getTaskName(), message.getTaskCount());
+
+    if (state.getWatermarkTime() != TIME_NOT_EXIST) {
+      long minTimestamp = watermarkStates.entrySet().stream()
+          .filter(entry -> entry.getKey().getSystemStream().equals(ssp.getSystemStream()))
+          .map(entry -> entry.getValue().getWatermarkTime())
+          .min(Long::compare)
+          .get();
+      Long curWatermark = watermarkPerStream.get(ssp.getSystemStream());
+      if (curWatermark == null || curWatermark < minTimestamp) {
+        watermarkPerStream.put(ssp.getSystemStream(), minTimestamp);
+        return new WatermarkImpl(minTimestamp);
+      }
+    }
+
+    return null;
+  }
+
+  /* package private */
+  long getWatermarkTime(SystemStreamPartition ssp) {
+    return watermarkStates.get(ssp).getWatermarkTime();
+  }
+
+  /**
+   * Send the watermark message to all partitions of an intermediate stream
+   * @param timestamp watermark timestamp
+   * @param systemStream intermediate stream
+   */
+  void sendWatermark(long timestamp, SystemStream systemStream, int taskCount) {
+    log.info("Send end-of-stream messages to all partitions of " + systemStream);
+    final WatermarkMessage watermarkMessage = new WatermarkMessage(timestamp, taskName, taskCount);
+    ControlMessageUtils.sendControlMessage(watermarkMessage, systemStream, metadataCache, collector);
+  }
+
+  /**
+   * Per ssp state of the watermarks. This class keeps track of the latest watermark timestamp
+   * from each upstream producer tasks, and use the min to update the aggregated watermark time.
+   */
+  final static class WatermarkState {
+    private int expectedTotal = Integer.MAX_VALUE;
+    private final Map<String, Long> timestamps = new HashMap<>();
+    private long watermarkTime = TIME_NOT_EXIST;
+
+    void update(long timestamp, String taskName, int taskCount) {
+      if (taskName != null) {
+        timestamps.put(taskName, timestamp);
+      }
+      expectedTotal = taskCount;
+
+      if (timestamps.size() == expectedTotal) {
+        Optional<Long> min = timestamps.values().stream().min(Long::compare);
+        watermarkTime = min.orElse(timestamp);
+      }
+    }
+
+    long getWatermarkTime() {
+      return watermarkTime;
+    }
+  }
+
+  /**
+   * Implementation of the Watermark. It keeps a reference to the {@link WatermarkManager}
```

Comment:
There should be some way for the user to control watermark emission (its fine, even good, if we do the right thing for the basic cases). The case where I need to control emission is when I'm pushing back work and then processing it based not on a watermark event but on a side input event (this would come through the normal event flow in Samza) - see DoFnOp from the Samza Beam runner as an example.

The WatermarkManager design is separating the handling of watermarks from the code that knows best how to handle watermarks (the operators). As a contrast, Flink has a cleaner design where watermark handling is baked into the operators with reasonable defaults and the user is able to entirely control watermark emission via an overridden operator. See AbstractStreamOperator where the default is to propagate watermarks, but the user can override processWatermark to change this. They can also emit watermarks from processWatermark or processeElement.

You can continue down the current design, but it is going to be more convoluted than Flink's and does not appear to have a compensating upside. If it's not obvious, please try to port DoFnOp to the proposed Samza API - it should make the design differences pretty clear. I suggest addressing user control of watermark emission now because I expect it will significantly change the user facing API and the implementation.

Happy to discuss further here or over video, but the port of DoFnOp should also be highly illuminating.